library/core/src/random.rs

```diff
@@ -0,0 +1,88 @@
+//! Random value generation.
+//!
+//! The [`Random`] trait allows generating a random value for a type using a
+//! given [`RandomSource`].
+
+/// A source of randomness.
+#[unstable(feature = "random", issue = "none")]
+pub trait RandomSource {
+    /// Fills `bytes` with random bytes.
+    fn fill_bytes(&mut self, bytes: &mut [u8]);
+}
+
+/// A trait for getting a random value for a type.
+///
+/// **Warning:** Be careful when manipulating random values! The
+/// [`random`](Random::random) method on integers samples them with a uniform
+/// distribution, so a value of 1 is just as likely as [`i32::MAX`]. By using
+/// modulo operations, some of the resulting values can become more likely than
+/// others. Use audited crates when in doubt.
+#[unstable(feature = "random", issue = "none")]
+pub trait Random: Sized {
+    /// Generates a random value.
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self;
+}
+
+impl Random for bool {
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
+        u8::random(source) & 1 == 1
+    }
+}
+
+impl Random for u8 {
+    /// Generates a random value.
+    ///
+    /// **Warning:** Be careful when manipulating the resulting value! This
+    /// method samples according to a uniform distribution, so a value of 1 is
+    /// just as likely as [`MAX`](Self::MAX). By using modulo operations, some
+    /// values can become more likely than others. Use audited crates when in
+    /// doubt.
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
+        let mut byte = [0];
+        source.fill_bytes(&mut byte);
+        byte[0]
+    }
+}
+
+impl Random for i8 {
+    /// Generates a random value.
+    ///
+    /// **Warning:** Be careful when manipulating the resulting value! This
+    /// method samples according to a uniform distribution, so a value of 1 is
+    /// just as likely as [`MAX`](Self::MAX). By using modulo operations, some
+    /// values can become more likely than others. Use audited crates when in
+    /// doubt.
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
+        u8::random(source) as i8
+    }
+}
```

Comment:
Just scrolled by here and I'm curious. Why are `u8` and `i8` not also implemented with the macro below? They also have `to_ne_bytes` and `from_ne_bytes` so might work nicely.